presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java

```diff
@@ -304,128 +290,146 @@ public void addInput(Page page)
         }
 
         pendingInput = page;
-        if (processPendingInput()) {
-            state = State.HAS_OUTPUT;
-        }
-        localUserMemoryContext.setBytes(pagesIndex.getEstimatedSize().toBytes());
     }
 
-    /**
-     * @return true if a full group has been buffered after processing the pendingInput, false otherwise
-     */
-    private boolean processPendingInput()
+    @Override
+    public Page getOutput()
     {
-        checkState(pendingInput != null);
-        pendingInput = updatePagesIndex(pendingInput);
-
-        // If we have unused input or are finishing, then we have buffered a full group
-        if (pendingInput != null || state == State.FINISHING) {
-            finishPagesIndex();
-            return true;
+        if (!outputPages.process()) {
+            return null;
         }
-        else {
-            return false;
+
+        if (outputPages.isFinished()) {
+            return null;
         }
+
+        return outputPages.getResult();
     }
 
-    /**
-     * @return the unused section of the page, or null if fully applied.
-     * pagesIndex guaranteed to have at least one row after this method returns
-     */
-    private Page updatePagesIndex(Page page)
+    @Override
+    public void close()
     {
-        checkArgument(page.getPositionCount() > 0);
+        driverWindowInfo.set(Optional.of(windowInfo.build()));
+    }
 
-        // TODO: Fix pagesHashStrategy to allow specifying channels for comparison, it currently requires us to rearrange the right side blocks in consecutive channel order
-        Page preGroupedPage = rearrangePage(page, preGroupedChannels);
-        if (pagesIndex.getPositionCount() == 0 || pagesIndex.positionEqualsRow(preGroupedPartitionHashStrategy, 0, 0, preGroupedPage)) {
-            // Find the position where the pre-grouped columns change
-            int groupEnd = findGroupEnd(preGroupedPage, preGroupedPartitionHashStrategy, 0);
+    private class ProduceWindowResults
+            implements WorkProcessor.Transformation<WindowPartition, Page>
+    {
+        private final PageBuilder pageBuilder;
 
-            // Add the section of the page that contains values for the current group
-            pagesIndex.addPage(page.getRegion(0, groupEnd));
+        private ProduceWindowResults()
+        {
+            pageBuilder = new PageBuilder(outputTypes);
```

Comment:
Sometimes reusing PageBuilder is valuable (because it can predict allocations' size better?)
Can we safely assume this isn't a case here?

Reply:
It is reused here, because `ProduceWindowResults` is long living